src/docgen/render/markdown.ts

```diff
@@ -0,0 +1,209 @@
+}
+
+/**
+ * Options for defining a markdown element.
+ */
+export interface MarkdownOptions {
+  /**
+   * Markdown header.
+   *
+   * @default - No header.
+   */
+  readonly header?: MarkdownHeaderOptions;
+
+  /**
+   * Id of the element.
+   *
+   * @default - The title will be used.
+   */
+  readonly id?: string;
+}
+
+/**
+ * Markdown element.
+ */
+export class Markdown {
+  /**
+   * An empty markdown element.
+   */
+  public static readonly EMPTY = new Markdown();
+
+  /**
+   * Sanitize markdown reserved characters from external input.
+   */
+  public static sanitize(line: string): string {
+    let sanitized = line.trim();
+
+    if (line.startsWith('-')) {
+      sanitized = sanitized.substring(1, line.length).trim();
+    }
+
+    return sanitized;
+  }
+
+  public static bold(text: string): string {
+    return `**${text}**`;
+  }
+
+  public static pre(text: string): string {
+    return `\`${text}\``;
+  }
+
+  public static italic(text: string) {
+    return `*${text}*`;
+  }
+
+  private readonly _lines = new Array<string>();
+  private readonly _sections = new Array<Markdown>();
+
+  private readonly id?: string;
+  private readonly header?: string;
+
+  constructor(private readonly options: MarkdownOptions = {}) {
+    this.id = options.id ?? options.header?.title;
+    this.header = this.formatHeader();
+  }
+
+  /**
+   * Render a `jsii-reflect.Docs` element into the markdown.
+   */
+  public docs(docs: reflect.Docs) {
+    if (docs.summary) {
+      this.lines(Markdown.sanitize(docs.summary));
+      this.lines('');
+    }
+    if (docs.remarks) {
+      this.lines(Markdown.sanitize(docs.remarks));
+      this.lines('');
+    }
+
+    if (docs.docs.see) {
+      this.quote(docs.docs.see);
+    }
+
+    const customLink = docs.customTag('link');
+    if (customLink) {
+      this.quote(`[${customLink}](${customLink})`);
+    }
+  }
+
+  public quote(line: string) {
+    this.lines(`> ${line}`);
+    this.lines('');
+  }
+
+  public bullet(line: string) {
+    this.lines(`- ${line}`);
+  }
+
+  public code(language: string, ...snippet: string[]) {
+    this.lines(`\`\`\`${language}`, ...snippet, '```');
+    this.lines('');
+  }
+
+  public lines(...lines: string[]) {
+    this._lines.push(...lines);
+  }
+
+  public split() {
+    this.lines('---');
+    this.lines('');
+  }
+
+  public section(section: Markdown) {
+    this._sections.push(section);
+  }
+
+  public render(headerSize: number = 0): string {
+    if (headerSize > 6) {
+      // headers are mapped to `h1-h6` html elements.
+      // passed that, markdown just renders `#` signs.
+      // lets see if and when we'll hit this limit.
+      throw new Error('Unable to render markdown. Header limit (6) reached.');
+    }
+
+    const content: string[] = [];
+    if (this.header) {
+      const anchor = sanitize(this.id ?? '');
+      const heading = `${'#'.repeat(headerSize)} ${this.header}`;
+      const headerSpan = !!process.env.HEADER_SPAN;
```

Comment:
This is nasty, i'm aware. 

Its just an escape hatch so that produce working links by default, but also support producing the links that `construct-hub` currently [relies](https://github.com/cdklabs/construct-hub-webapp/blob/main/src/components/PackageDocs/PackageDocs.tsx#L86) on.

This will be gone soon.